articles/realtime.md

```diff
@@ -0,0 +1,259 @@
+
+Real-time software guarantees correct computation at the correct time.
+Hard real-time software systems have a set of strict deadlines, and missing a deadline is considered a failure.
+Soft real-time systems degrade their quality of service if a deadline is missed.
+
+Real-time computer systems are often associated with low-latency systems.
+Many applications of real-time computing are also low-latency applications (for example, automated piloting systems must be reactive to sudden changes in the environment).
+However, it is generally agreed upon that a real-time system is not defined by low latency, but by a deterministic schedule: it must be guaranteed that the system finishes a certain task by a certain time.
+Therefore, it is important that the latency in the system be measurable and a maximum allowable latency for tasks be set.
+
+A more useful metric in evaluating the "hardness" of a real-time system is jitter, which is defined as the variable deviation between a task's deadline and its actual time of completion.
+A more jittery system is less deterministic, less predictable, and less real-time.
+Though in practice it is impossible to completely eliminate jitter from a real-time system, it is a worthy goal to determine a hard upper bound for jitter.
+
+# Implementation of Real-time Computing
+
+In general, an operating system can guarantee that the tasks it handles for the developer, such as thread scheduling, are deterministic, but the OS may not guarantee that the developer's code will run in real-time.
+Therefore, it is up to the developer to know what the determinstic guarantees of an existing system are, and what she must do to write hard real-time code on top of the OS.
+
+In this section, various strategies for developing on top of a real-time OS are explored, since these strategies might be applicable to ROS 2.
+The patterns focus on the use case of C/C++ development on Linux-based real-time OS's (such as RTLinux), but the general concepts are applicable to other platforms.
+
+## Memory management
+
+### Lock memory, prefault stack:
+
+```c
+    if (mlockall(MCL_CURRENT|MCL_FUTURE) == -1) {
+            perror("mlockall failed");
+            exit(-2);
+    }
+    unsigned char dummy[MAX_SAFE_STACK];
+
+    memset(dummy, 0, MAX_SAFE_STACK);
+```
+
+`mlockall` is a Linux system call for locking the process's virtual address space into RAM, preventing the memory that will be accessed by the process from getting paged into swap space.
+Paging and accessing paged memory is a nondeterministic operation that should be avoided in real-time computation.
+
+This code snippet, when run at the beginning of a thread's lifecycle, ensures that no pagefaults occur while the thread is running.
+`mlockall` locks the stack for the thread.
+The `memset` call pre-loads each block of memory of the stack into the cache, so that no pagefaults will occur when the stack is accessed.
+
+### Allocate dynamic memory pool
```

Comment:
Before going into details of the solutions, mention what is the problem with dynamic memory allocation. This [Wikipedia](https://en.wikipedia.org/wiki/Real-time_operating_system#Memory_allocation) entry provides a good summary. It mentions issues with memory fragmentation, and the non-deterministic nature of standard memory allocators.

Generally speaking, I would recommend strategies that depend on the problem at hand:

* Data size is known at complile-time: Use fixed objects allocated on the stack (if reasonably sized).
* Data size is known at runtime, during configuration, before starting RT-scheduled execution. Pre-allocate resources on the heap. Node lifecycle becomes relevant here, as we can clearly state in which transitions a real-time safe node can perform non real-time safe calls.
* Data size is known at (or can change during) RT-scheduled execution. Multiple alternatives exist, which is mostly what you focus on below. Typical strategies are to use object pools (good if you have some context of the problem), general-purpose memory pools like the [TLSF](http://www.gii.upv.es/tlsf/) O(1) allocator, stack allocation, etc. In such cases there's typically a buffer that can wither get full or overflow, so AFAIK, there's no silver bullet.